finalfusion-ffi/src/vocab.rs

```diff
@@ -0,0 +1,70 @@
+use std::ffi::CStr;
+use std::mem;
+use std::os::raw::{c_char, c_int};
+
+use finalfusion::chunks::vocab::WordIndex;
+use finalfusion::prelude::*;
+
+use crate::check_null;
+
+/// Indices lookup.
+///
+/// Returns an integer indicating how a word is represented:
+///    * 0 if the word cannot be represented
+///    * `n` if the word is represented by `n` embeddings
+///
+/// Writes the indices to `indices_res` if it is not a null-pointer.
+#[no_mangle]
+pub unsafe extern "C" fn ff_vocab_indices(
+    embeddings: *mut Embeddings<VocabWrap, StorageWrap>,
+    word: *const c_char,
+    indices_res: *mut *mut usize,
+) -> c_int {
+    check_null!(embeddings);
+    check_null!(word);
+    let embeddings = &*embeddings;
+    let vocab = embeddings.vocab();
+    let word = CStr::from_ptr(word).to_string_lossy();
+    if let Some(idx) = vocab.idx(&word) {
```

Comment:
Maybe it's clearer when this whole block is outdented? Something like

```rust
let idx = match vocab.idx(&word) {
  Some(idx) => idx,
  None => return 0,
}
```

then the rest of the function can be `flattened` a bit.

Reply:
Also, I guess the other `match` block could be removed if we had an `indices` method on `WordIndex`. Your proposed iterator would offer the same, but then there would be a vec -> iter -> vec transition. `into_inner` would also be a possible name.

Reply:
valgrind reports `Conditional jump or move depends on uninitialised value(s)` if this is moved outside the match:
~~~Rust
            if !indices_res.is_null() {
                let ptr = indices.as_mut_ptr();
                mem::forget(indices);
                *indices_res = ptr;
            }
~~~

Any idea what's up with that?

Reply:
The fact that it works in debug but not in release hints at some optimization.

Maybe a single-value optimization for the `WordIndex::Word` case, where the value gets stored on the stack somehow? Best to look at the machine code if you dare ;).

Reply:
The `Vec` documentation states:

> Vec will never perform a "small optimization" where elements are actually stored on the stack for two reasons:

Of course, this could be something that LLVM did. Or I am overlooking something simple ;).

Reply:
Another problem may be that the compiler infers (somehow) that indices is not used anymore after the `match` scope. You could try to add a `mem::drop(indices)` in an `else` to the `if` to compel the compiler to believe that it is not dropped.

Will be interesting to see what the error is, and if this is some compiler error.

Reply:
Ok, has nothing to do with the `forget`.

Reply:
Ah, doh, it's the `is_null`.

Reply:
Sure, it's a conditional based on whether something is null, but that's perfectly fine and common, no?

Reply:
No, but it seems that `valgrind` believes that the pointer is uninitialized memory:

```
==20605== Conditional jump or move depends on uninitialised value(s)
==20605==    at 0x485D922: ff_vocab_indices (vocab.rs:46)
==20605==    by 0x4010DA: main (vocab_test.c:14)
==20605==  Uninitialised value was created by a stack allocation
==20605==    at 0x401070: ??? (in /home/daniel/git/finalfusion-ffi/build/tests/vocab_test)
```

And of course, branching on uninitialized memory would be illegal. But I don't see why that's the case:

```
(gdb) info args
embeddings = 0x4083e0
word = 0x402017 "Berlin"
indices_res = 0x7fffffff66b0
(gdb) up
(gdb) print &indices
$7 = (size_t **) 0x7fffffff66b0
```

So, `indices_res` it seems to be a correct, valid pointer. Also, the memory address seems to be correct, it's near the start of the frame (remember stack grows upward).

```
(gdb) info f
Stack level 1, frame at 0x7fffffff66d0:
 rip = 0x4010db in main (/home/daniel/git/finalfusion-ffi/tests/vocab_test.c:14); 
    saved rip = 0x7ffff7da3b8e
 caller of frame at 0x7fffffff66b0
 source language c.
 Arglist at 0x7fffffff66a8, args: 
 Locals at 0x7fffffff66a8, Previous frame's sp is 0x7fffffff66d0
 Saved registers:
  rbx at 0x7fffffff66c0, rip at 0x7fffffff66c8
```

Reply:
Wrapping with defuses the warning since it explicitly checks for null I guess?
~~~Rust
    if !indices_res.is_null() {
        let ptr = indices.as_mut_ptr();
        mem::forget(indices);
        *indices_res = ptr;
    }
~~~

Reply:
I stepped through the execution with gdb and for some reason, it is dropping a `RawVec`. Since nearly everything is optimized out, I can't really see what pointer that `RawVec` is holding and if it actually deallocates it.

Reply:
I'm not sure what the implications of this are, it doesn't look like a programming error from our side but we can circumvent it by writing the method differently.

But it might be worthy to open an issue at the Rust repo?